### ISSUE TITLE
Cannot run Orf Inference

### ISSUE THREAD

**Issue by yeredh** [erhard-lab/gedi#41]
Hello Florian,

I downloaded the FASTQ files for the Ribo-seq experiments in GSE114636 from ENA (https://www.ebi.ac.uk/ena), and I used `gedi -e Pipeline` to generate all the Price scripts. The scripts fail in the Orf inference step and I get the following error message

```
2020-03-23 09:44:51.691 INFO OS: Linux 3.10.0-862.el7.x86_64 amd64
2020-03-23 09:44:51.741 INFO Java: OpenJDK 64-Bit Server VM 1.8.0_232-b09
2020-03-23 09:44:51.795 INFO Discovering classes in classpath
2020-03-23 09:44:51.833 INFO Preparing simple class references
2020-03-23 09:44:51.887 INFO Gedi version 1.0.2 (JAR) startup
2020-03-23 09:44:51.889 INFO Command: gedi -e Price -reads /data/work/Projects/RiboSeq/GSE114636/./GSE114636_PRICE/GSE114636_PRICE.cit -genomic GRCh38 -prefix /data/work/Projects/RiboSeq/GSE114636/./GSE114636_PRICE/price/GSE114636_PRICE -plot -percond -D
2020-03-23 09:44:51.889 INFO Price version 1.0.3b
2020-03-23 09:44:51.941 SEVERE Could not load startup FxStartup.class in /data/resources/tools/Price_1.0.3b/gedi.jar
2020-03-23 09:44:52.011 INFO Reading oml /home/yered/.gedi/genomic/GRCh38.oml
2020-03-23 09:44:52.015 INFO Done reading oml /home/yered/.gedi/genomic/GRCh38.oml
2020-03-23 09:44:52.797 INFO Collect signal to noise data
2020-03-23 09:44:53.617 INFO Running R scripts for plotting
2020-03-23 09:44:54.281 INFO Infer ORFs
PRICE
PRICE is an analysis method for Ribo-seq data.

Error:
java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@567d4d32
Could not run gedi.riboseq.javapipeline.PriceOrfInference@567d4d32
Exception in parallel iterator thread!
0

gedi -e PRICE <Options>

General:
 -reads <reads>        The mapped reads from the ribo-seq experiment.
 -prefix <prefix>      The prefix used for all output files
 -genomic <genomic>    The indexed GEDI genome.

Commandline:
 -progress             Show progress
 -D                    Verbose output of errors
 -h                    Show usage
 -hh                   Show verbose usage
 -hhh                  Show extra verbose usage
 -dry                  Dry run
 -keep                 Do not remove temp files


java.lang.RuntimeException: java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@567d4d32
	at gedi.util.job.schedule.DefaultPetriNetScheduler.run(DefaultPetriNetScheduler.java:223)
	at gedi.util.program.GediProgram$1.execute(GediProgram.java:375)
	at gedi.util.program.GediProgram.run(GediProgram.java:210)
	at executables.Price.main(Price.java:70)
Caused by: java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@567d4d32
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:63)
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:27)
	at gedi.util.job.FireTransition.call(FireTransition.java:54)
	at gedi.util.job.FireTransition.call(FireTransition.java:25)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Exception in parallel iterator thread!
	at gedi.util.functions.ParallelizedIterator.fatal(ParallelizedIterator.java:242)
	at gedi.util.functions.ParallelizedIterator.tryNext(ParallelizedIterator.java:209)
	at gedi.util.functions.ParallelizedIterator.hasNext(ParallelizedIterator.java:170)
	at java.util.Iterator.forEachRemaining(Iterator.java:115)
	at gedi.riboseq.utils.RiboUtils.lambda$processCodonsSink$4(RiboUtils.java:107)
	at gedi.riboseq.utils.RiboUtils.processCodons(RiboUtils.java:147)
	at gedi.riboseq.utils.RiboUtils.processCodonsSink(RiboUtils.java:107)
	at gedi.riboseq.javapipeline.PriceOrfInference.execute(PriceOrfInference.java:118)
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:61)
	... 7 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at jdistlib.math.spline.SmoothSpline.findDuplicateIndices(SmoothSpline.java:1508)
	at jdistlib.math.spline.SmoothSpline.fit(SmoothSpline.java:1392)
	at jdistlib.math.spline.SmoothSpline.fitDFMatch(SmoothSpline.java:1296)
	at gedi.riboseq.inference.orf.NoiseModel.computeMeanSpline(NoiseModel.java:188)
	at gedi.riboseq.inference.orf.NoiseModel.getProbability(NoiseModel.java:141)
	at gedi.riboseq.inference.orf.NoiseModel.getProbability0(NoiseModel.java:121)
	at gedi.riboseq.inference.orf.OrfInference.inFrameTestCandidate(OrfInference.java:1272)
	at gedi.riboseq.inference.orf.OrfInference.inFrameTest(OrfInference.java:1080)
	at gedi.riboseq.inference.orf.OrfInference.inferOrfs(OrfInference.java:434)
	at gedi.riboseq.javapipeline.PriceOrfInference.lambda$null$1(PriceOrfInference.java:121)
	at gedi.util.FunctorUtils$DemultiplexIterator.lookAhead(FunctorUtils.java:108)
	at gedi.util.FunctorUtils$DemultiplexIterator.hasNext(FunctorUtils.java:92)
	at gedi.util.FunctorUtils$PeekIterator.hasNext(FunctorUtils.java:640)
	at gedi.util.FunctorUtils$SideEffectIterator.hasNext(FunctorUtils.java:721)
	at gedi.util.functions.ExtendedIterator.toCollection(ExtendedIterator.java:450)
	at gedi.util.functions.ParallelizedIterator$Worker.run(ParallelizedIterator.java:360)
2020-03-23 09:44:54.461 INFO Finished: gedi -e Price -reads /data/work/Projects/RiboSeq/GSE114636/./GSE114636_PRICE/GSE114636_PRICE.cit -genomic GRCh38 -prefix /data/work/Projects/RiboSeq/GSE114636/./GSE114636_PRICE/price/GSE114636_PRICE -plot -percond -D
2020-03-23 09:44:54.462 INFO Took 2s 777ms
Warning message:
Removed 475 rows containing missing values (geom_label). 
```
The JSON file that I used to generate the scripts was
```
{
"datasets": [
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182546.fastq.gz", "name": "cyclo_untr", "adapter": "ATTGATCATCGA"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182547.fastq.gz", "name": "cyclo_ifn", "adapter": "ATTGATCATCGAC"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182548.fastq.gz", "name": "cyclo_vir", "adapter": "ATTGATCATCGA"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182549.fastq.gz", "name": "cyclo_ifn_vir", "adapter": "ATTGATCATCGAC"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182550.fastq.gz", "name": "ltm_untr", "adapter": "ATTGATCATCGAC"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182551.fastq.gz", "name": "ltm_ifn", "adapter": "ATTGATCATCGAC"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182552.fastq.gz", "name": "ltm_vir", "adapter": "ATTGATCATCGAC"},
		{"fastq": "/data/work/Projects/RiboSeq/GSE114636/fastq/SRR7182553.fastq.gz", "name": "ltm_ifn_vir", "adapter": "ATTGATCATCGA"}
	],
"references": {"GRCh38": "both"}
}
```

**Comment by jflucier** [comment]
Hi Florian,

I have exactly the same problem as yeredh on our system:


2020-05-14 10:29:18.129 INFO Estimate parameters
Correcting from 6 to 30 (desired: 31)
2020-05-14 10:32:27.550 INFO LL=-1.00572e+08
2020-05-14 10:32:27.551 INFO Using maxpos=31 (ERX609917)
2020-05-14 10:32:27.552 INFO Estimate parameters
Correcting from 9 to 30 (desired: 31)
2020-05-14 10:36:35.017 INFO LL=-1.47656e+07
2020-05-14 10:36:35.019 INFO Using maxpos=31 (ERX609918)
2020-05-14 10:36:35.019 INFO Estimate parameters
Correcting from 9 to 30 (desired: 31)
2020-05-14 10:39:26.448 INFO LL=-1.31339e+07
2020-05-14 10:39:26.450 INFO Using maxpos=31 (ERX609919)
2020-05-14 10:39:26.450 INFO Estimate parameters
Correcting from 12 to 30 (desired: 31)
2020-05-14 10:43:13.709 INFO LL=-7.17452e+07
2020-05-14 10:43:13.711 INFO Using maxpos=31 (ERX609920)
2020-05-14 10:43:13.712 INFO Estimate parameters
Correcting from 12 to 30 (desired: 31)
2020-05-14 10:47:05.398 INFO LL=-1.03898e+08
2020-05-14 10:47:05.401 INFO Using maxpos=12 (ERX609921)
2020-05-14 10:47:05.401 INFO Estimate parameters
2020-05-14 10:48:16.322 INFO LL=-5.21327e+07
2020-05-14 10:48:16.324 INFO Using maxpos=12 (ERX609922)
2020-05-14 10:48:16.324 INFO Estimate parameters
2020-05-14 10:50:14.180 INFO LL=-1.49199e+07
2020-05-14 10:50:14.181 INFO Using maxpos=12 (ERX609923)
2020-05-14 10:50:14.182 INFO Estimate parameters
2020-05-14 10:51:57.313 INFO LL=-5.81962e+07
2020-05-14 10:51:57.316 INFO Using maxpos=12 (ERX609924)
2020-05-14 10:51:57.317 INFO Estimate parameters
2020-05-14 10:53:34.738 INFO LL=-6.35817e+07
2020-05-14 10:53:34.739 INFO Using maxpos=12 (ERX609925)
2020-05-14 10:53:34.740 INFO Estimate parameters
2020-05-14 10:54:48.008 INFO LL=-6.61868e+07
2020-05-14 10:54:48.107 INFO Codon inference
2020-05-14 11:30:53.842 INFO Train start prediction
2020-05-14 11:30:53.843 INFO Calibrate noise model
2020-05-14 11:30:53.843 INFO Writing codon index
2020-05-14 11:42:47.272 INFO Considering 1 ORFs for start training
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: initial function value: 4.1589
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: the function value after   5 iterations: 2.7034
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: initial function value: 4.1589
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: the function value after  10 iterations: 2.5129
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: the function value after  20 iterations: 2.5020
[GediProgram-1-thread-10] INFO smile.math.Math - L-BFGS: the function value after  29 iterations: 2.5020
2020-05-14 11:42:47.363 INFO Infer ORFs
PRICE
PRICE is an analysis method for Ribo-seq data.

Error:
java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@6247f7b5
Could not run gedi.riboseq.javapipeline.PriceOrfInference@6247f7b5
Exception in parallel iterator thread!
0

gedi -e PRICE <Options>

General:
 -reads <reads>        The mapped reads from the ribo-seq experiment.
 -prefix <prefix>      The prefix used for all output files
 -genomic <genomic>    The indexed GEDI genome.

Commandline:
 -progress             Show progress
 -D                    Verbose output of errors
 -h                    Show usage
 -hh                   Show verbose usage
 -hhh                  Show extra verbose usage
 -dry                  Dry run
 -keep                 Do not remove temp files


java.lang.RuntimeException: java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@6247f7b5
	at gedi.util.job.schedule.DefaultPetriNetScheduler.run(DefaultPetriNetScheduler.java:223)
	at gedi.util.program.GediProgram$1.execute(GediProgram.java:359)
	at gedi.util.program.GediProgram.run(GediProgram.java:204)
	at executables.Price.main(Price.java:68)
Caused by: java.lang.RuntimeException: Could not run gedi.riboseq.javapipeline.PriceOrfInference@6247f7b5
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:63)
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:27)
	at gedi.util.job.FireTransition.call(FireTransition.java:54)
	at gedi.util.job.FireTransition.call(FireTransition.java:25)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Exception in parallel iterator thread!
	at gedi.util.functions.ParallelizedIterator.fatal(ParallelizedIterator.java:221)
	at gedi.util.functions.ParallelizedIterator.tryNext(ParallelizedIterator.java:189)
	at gedi.util.functions.ParallelizedIterator.hasNext(ParallelizedIterator.java:150)
	at java.util.Iterator.forEachRemaining(Iterator.java:115)
	at gedi.riboseq.utils.RiboUtils.lambda$processCodonsSink$4(RiboUtils.java:106)
	at gedi.riboseq.utils.RiboUtils.processCodons(RiboUtils.java:146)
	at gedi.riboseq.utils.RiboUtils.processCodonsSink(RiboUtils.java:106)
	at gedi.riboseq.javapipeline.PriceOrfInference.execute(PriceOrfInference.java:118)
	at gedi.util.program.GediProgramJob.execute(GediProgramJob.java:61)
	... 7 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at jdistlib.math.spline.SmoothSpline.findDuplicateIndices(SmoothSpline.java:1508)
	at jdistlib.math.spline.SmoothSpline.fit(SmoothSpline.java:1392)
	at jdistlib.math.spline.SmoothSpline.fitDFMatch(SmoothSpline.java:1296)
	at gedi.riboseq.inference.orf.NoiseModel.computeMeanSpline(NoiseModel.java:188)
	at gedi.riboseq.inference.orf.NoiseModel.getProbability(NoiseModel.java:141)
	at gedi.riboseq.inference.orf.NoiseModel.getProbability0(NoiseModel.java:121)
	at gedi.riboseq.inference.orf.OrfInference.inFrameTestCandidate(OrfInference.java:1271)
	at gedi.riboseq.inference.orf.OrfInference.inFrameTest(OrfInference.java:1079)
	at gedi.riboseq.inference.orf.OrfInference.inferOrfs(OrfInference.java:433)
	at gedi.riboseq.javapipeline.PriceOrfInference.lambda$null$1(PriceOrfInference.java:119)
	at gedi.util.FunctorUtils$DemultiplexIterator.lookAhead(FunctorUtils.java:106)
	at gedi.util.FunctorUtils$DemultiplexIterator.hasNext(FunctorUtils.java:90)
	at gedi.util.FunctorUtils$PeekIterator.hasNext(FunctorUtils.java:636)
	at gedi.util.FunctorUtils$SideEffectIterator.hasNext(FunctorUtils.java:717)
	at gedi.util.functions.ExtendedIterator.toCollection(ExtendedIterator.java:375)
	at gedi.util.functions.ParallelizedIterator$Worker.run(ParallelizedIterator.java:333)
2020-05-14 11:42:48.076 INFO Finished: gedi -e Price -nthreads 23 -reads /ip29/jflucier/temp/ERP007231/ERP007231.cit -genomic rat_rrna rat_ensembl -prefix /ip29/jflucier/temp/ERP007231/price_fdr_1per/ERP007231 -percond -D -fdr 0.01
|11:42:50|jflucier@ip29:[ERP007231]> 


tks for your help,
JF

**Comment by florianerhard** [comment]
Dear both,

these kinds of errors most likely are the consequence of read mapping problems. In case of the flu data, I just checked: They used a library preparation that is completely new to me (poly-adenylation + circularization instead of adapter ligation). This has two consequences: If you let the pipeline trim the given adapter (where did you get that sequence?), most of the reads will not be trimmed and contain long polyA stretches, and only a minuscule amount of reads will be mapped. The other consequence is that if you trim polyA at the 3' end, you might trim too many A (i.e. there might be one or several templated A that are trimmed). This is unexpected and not modeled in price, which make subcodon resolution analyses difficult...

Best,
Florian

**Comment by yeredh** [comment]
Dear Florian,

Thank you very much for taking the time to reply. I was not able to run the ORF inference, because the mapping rate was terrible. The low mapping rate was because I was trying to infer the adapter with DNApi. I was getting the wrong adapter, and trimming the sequences incorrectly. Once I provided the right sequence (AAAA...AAAA) for the adapter, I was able to run the ORF inference successfully.

Best,

Yered
